cpp/src/arrow/stl_test.cc

```diff
@@ -501,4 +504,59 @@ TEST(TestTupleVectorFromTable, CastingNeeded) {
 }
 
 }  // namespace stl
+
+TEST(STLMemoryPool, Base) {
+  std::allocator<uint8_t> allocator;
+  STLMemoryPool<std::allocator<uint8_t>> pool(allocator);
+
+  uint8_t* data = nullptr;
+  ASSERT_OK(pool.Allocate(100, &data));
+  ASSERT_EQ(pool.max_memory(), 100);
+  ASSERT_EQ(pool.bytes_allocated(), 100);
+  ASSERT_NE(data, nullptr);
+
+  ASSERT_OK(pool.Reallocate(100, 150, &data));
+  ASSERT_EQ(pool.max_memory(), 150);
+  ASSERT_EQ(pool.bytes_allocated(), 150);
+
+  pool.Free(data, 150);
+
+  ASSERT_EQ(pool.max_memory(), 150);
+  ASSERT_EQ(pool.bytes_allocated(), 0);
+}
+
+TEST(stl_allocator, MemoryTracking) {
+  auto pool = default_memory_pool();
+  stl_allocator<uint64_t> alloc;
+  uint64_t* data = alloc.allocate(100);
+
+  ASSERT_EQ(100 * sizeof(uint64_t), pool->bytes_allocated());
+
+  alloc.deallocate(data, 100);
+  ASSERT_EQ(0, pool->bytes_allocated());
+}
+
+#if !(defined(ARROW_VALGRIND) || defined(ADDRESS_SANITIZER) || defined(ARROW_JEMALLOC))
+
+TEST(stl_allocator, TestOOM) {
+  stl_allocator<uint64_t> alloc;
+  uint64_t to_alloc = std::numeric_limits<uint64_t>::max() / 2;
+  ASSERT_THROW(alloc.allocate(to_alloc), std::bad_alloc);
+}
+
+TEST(stl_allocator, MaxMemory) {
+  auto pool = default_memory_pool();
```

Comment:
`default_memory_pool()` returns a pointer to a singleton, which I suspect is what's [causing this test to fail](https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/27814557/job/vrxo430wmmptaafp#L1637). I believe you can fix it by using `MemoryPool::CreateDefault()` (which explicitly creates a new instance)